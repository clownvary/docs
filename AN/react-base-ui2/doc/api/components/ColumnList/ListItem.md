## ListItem 



### Prop Types
Property | Type | Default Value | Is Required | Description
:--- | :--- | :--- | :--- | :---
item|object|{}|false|&ensp;
index|number|0|false|&ensp;
disabled|bool|false|false|&ensp;
showTips|bool|false|false|&ensp;
selected|bool|false|false|&ensp;
focused|bool|false|false|&ensp;
onFocus|func|identity|false|&ensp;
onClick|func|identity|false|&ensp;
onRender|func|null|false|&ensp;
columns|[{<br>&emsp;&emsp;field: string,<br>&emsp;&emsp;type: string,<br>&emsp;&emsp;onRender: func<br>}]|[]|false|&ensp;