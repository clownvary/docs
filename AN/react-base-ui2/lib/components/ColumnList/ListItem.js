'use strict';Object.defineProperty(exports, "__esModule", { value: true });var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);var _createClass2 = require('babel-runtime/helpers/createClass');var _createClass3 = _interopRequireDefault(_createClass2);var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);var _inherits2 = require('babel-runtime/helpers/inherits');var _inherits3 = _interopRequireDefault(_inherits2);var _react = require('react');var _react2 = _interopRequireDefault(_react);
var _propTypes = require('prop-types');var _propTypes2 = _interopRequireDefault(_propTypes);
var _classnames = require('classnames');var _classnames2 = _interopRequireDefault(_classnames);
var _identity = require('lodash/identity');var _identity2 = _interopRequireDefault(_identity);
var _isFunction = require('lodash/isFunction');var _isFunction2 = _interopRequireDefault(_isFunction);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}

var ListItemPropTypes = {
  item: _propTypes2.default.object,
  index: _propTypes2.default.number,
  disabled: _propTypes2.default.bool,
  showTips: _propTypes2.default.bool,
  selected: _propTypes2.default.bool,
  focused: _propTypes2.default.bool,
  onFocus: _propTypes2.default.func,
  onClick: _propTypes2.default.func,
  onRender: _propTypes2.default.func,
  columns: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    field: _propTypes2.default.string.isRequired,
    type: _propTypes2.default.string.isRequired,
    onRender: _propTypes2.default.func })) };



var ListItemProps = {
  item: {},
  index: 0,
  focused: false,
  selected: false,
  disabled: false,
  showTips: false,
  onFocus: _identity2.default,
  onClick: _identity2.default,
  onRender: null,
  columns: [] };var


ListItem = function (_React$PureComponent) {(0, _inherits3.default)(ListItem, _React$PureComponent);function ListItem() {var _ref;var _temp, _this, _ret;(0, _classCallCheck3.default)(this, ListItem);for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {args[_key] = arguments[_key];}return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = ListItem.__proto__ || (0, _getPrototypeOf2.default)(ListItem)).call.apply(_ref, [this].concat(args))), _this), _this.




    onFocus = function (e) {var _this$props =
      _this.props,index = _this$props.index,focused = _this$props.focused,onFocus = _this$props.onFocus;
      focused || onFocus(e, { index: index });
    }, _this.

    onClick = function (e) {var _this$props2 =
      _this.props,item = _this$props2.item,index = _this$props2.index,onClick = _this$props2.onClick;
      onClick(e, { item: item, index: index });
    }, _this.

    renderItem = function (item, itemProps, context) {return (
        context.columns.map(function (column, columnIndex) {
          var value = item[column.field];
          var props = {
            key: columnIndex,
            disabled: itemProps.disabled,
            className: (0, _classnames2.default)('item-column', column.className) };

          context.column = column;
          context.columnIndex = columnIndex;
          return column.render(value, props, context);
        }));}, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);}(0, _createClass3.default)(ListItem, [{ key: 'render', value: function render()

    {var _props =









      this.props,item = _props.item,index = _props.index,columns = _props.columns,showTips = _props.showTips,focused = _props.focused,selected = _props.selected,disabled = _props.disabled,onRender = _props.onRender;

      var itemProps = { disabled: disabled };
      var context = { item: item, index: index, columns: columns };

      return (
        _react2.default.createElement('li', {
            role: 'option',
            tabIndex: 0,
            onClick: this.onClick,
            onFocus: this.onFocus,
            'aria-selected': selected,
            title: showTips ? item.text : null,
            className: (0, _classnames2.default)('item', item.className, {
              disabled: disabled,
              selected: selected,
              focused: focused }) },



          (0, _isFunction2.default)(onRender) ?
          onRender(item, itemProps, context) :
          this.renderItem(item, itemProps, context)));



    } }]);return ListItem;}(_react2.default.PureComponent);ListItem.displayName = 'ListItem';ListItem.propTypes = ListItemPropTypes;ListItem.defaultProps = ListItemProps;exports.default = ListItem;module.exports = exports['default'];