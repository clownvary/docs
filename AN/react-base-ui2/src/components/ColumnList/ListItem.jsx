import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import identity from 'lodash/identity';
import isFunction from 'lodash/isFunction';

const ListItemPropTypes = {
  item: PropTypes.object,
  index: PropTypes.number,
  disabled: PropTypes.bool,
  showTips: PropTypes.bool,
  selected: PropTypes.bool,
  focused: PropTypes.bool,
  onFocus: PropTypes.func,
  onClick: PropTypes.func,
  onRender: PropTypes.func,
  columns: PropTypes.arrayOf(PropTypes.shape({
    field: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    onRender: PropTypes.func
  }))
};

const ListItemProps = {
  item: {},
  index: 0,
  focused: false,
  selected: false,
  disabled: false,
  showTips: false,
  onFocus: identity,
  onClick: identity,
  onRender: null,
  columns: []
};

export default class ListItem extends React.PureComponent {
  static displayName = 'ListItem';
  static propTypes = ListItemPropTypes;
  static defaultProps = ListItemProps;

  onFocus = (e) => {
    const { index, focused, onFocus } = this.props;
    focused || onFocus(e, { index });
  }

  onClick = (e) => {
    const { item, index, onClick } = this.props;
    onClick(e, { item, index });
  }

  renderItem = (item, itemProps, context) =>
    context.columns.map((column, columnIndex) => {
      const value = item[column.field];
      const props = {
        key: columnIndex,
        disabled: itemProps.disabled,
        className: classNames('item-column', column.className)
      };
      context.column = column;
      context.columnIndex = columnIndex;
      return column.render(value, props, context);
    })

  render() {
    const {
      item,
      index,
      columns,
      showTips,
      focused,
      selected,
      disabled,
      onRender
    } = this.props;

    const itemProps = { disabled };
    const context = { item, index, columns };

    return (
      <li
        role="option"
        tabIndex={0}
        onClick={this.onClick}
        onFocus={this.onFocus}
        aria-selected={selected}
        title={showTips ? item.text : null}
        className={classNames('item', item.className, {
          disabled,
          selected,
          focused
        })}
      >
        {
          isFunction(onRender)
            ? onRender(item, itemProps, context)
            : this.renderItem(item, itemProps, context)
        }
      </li>
    );
  }
}
