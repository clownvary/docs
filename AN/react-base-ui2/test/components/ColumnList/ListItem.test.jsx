import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import ListItem from 'src/components/ColumnList/ListItem';

describe('components/ColumnList/ListItem', () => {
  it('ListItem renders fine', () => {
    const columnRender = value => (<div key={value}>{value}</div>)
    const props = {
      item: {
        id: '1234',
        name: 'question-1'
      },
      index: 0,
      columns: [
        { field: 'id', className: 'item-id', type: 'text', render: columnRender },
        { field: 'name', className: 'item-name', type: 'text', render: columnRender }
      ],
      showTips: true,
      focused: true,
      selected: true,
      disabled: true
    };

    const snapshot = renderer.create(
      <ListItem {...props} />
    ).toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  it('ListItem component renders fine', () => {
    const props = {
      item: {
        id: '7712',
        name: 'answer-1',
        value: 'YES'
      },
      columns: [
        { field: 'id', type: 'text' },
        { field: 'name', type: 'text' },
        { field: 'value', type: 'text' }
      ],
      onRender: jest.fn(),
      onFocus: jest.fn(),
      onClick: jest.fn()
    };

    const component = shallow(<ListItem {...props} />);
    expect(props.onRender).toHaveBeenCalled();

    const li = component.find('li');
    expect(li).toBeTruthy();

    li.simulate('focus');
    expect(props.onFocus).toHaveBeenCalled();

    li.simulate('click');
    expect(props.onClick).toHaveBeenCalled();
  });
});
