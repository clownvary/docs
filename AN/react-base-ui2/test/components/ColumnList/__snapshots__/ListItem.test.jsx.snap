// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ColumnList/ListItem ListItem renders fine 1`] = `
<li
  aria-selected={true}
  className="item disabled selected focused"
  onClick={[Function]}
  onFocus={[Function]}
  role="option"
  tabIndex={0}
  title={undefined}
>
  <div>
    1234
  </div>
  <div>
    question-1
  </div>
</li>
`;
